modules/README.md

```diff
@@ -1,58 +1,164 @@
-                    "properties": {
-                        "name": {
-                            "type": "string"
-                        },
-                        "version": {
-                            "type": "string"
-                        },
+        "type" : "object",
+        "description" : "The action argument passed via stdin",
+        "properties" : {
+            "input" : {
+                "type" : "object",
+                "description" : "PXP request arguments (complies with action's input in metadata)"
+            },
+            "configuration" : {
+                "type" : "object",
+                "description" : "Module's configuration (complies with metadata's configuration)"
+            },
+            "output_files" : {
+                "type" : "object",
+                "description" : "Contains the paths of files for storing the action output",
+                "properties" : {
+                    "stdout" : {
+                        "type" : "string",
+                        "description" : "File for storing the output on stdout"
                     },
-                    "required": [ "name" ]
-                },
-                "output": {
-                    "type": "object",
-                    "properties": {
-                        "success": {
-                            "type": "boolean"
-                        }
+                    "stderr" : {
+                        "type" : "string",
+                        "description" : "File for storing the output on stderr"
                     },
-                    "required": [ "success" ]
-                }
+                    "exitcode" : {
+                        "type" : "string",
+                        "description" : "File for storing the action's exitcode"
+                    },
+                },
+                "required" : ["out", "err", "exitcode"],
+                "additionalProperties" : false,
             }
-         ]
-     }
+        },
+        "required" : ["input"],
+        "additionalProperties" : false,
+    }
+
+The `input` and `configuration` entries of an input object comply with the
+respective schemas included in the `metadata`; pxp-agent enforces that. Note
+that the `configuration` schema is optional; in case it's not included in the
+`metadata`, the content of the configuration file will be included in the input
+object if its format is valid JSON.
 
+Each action should write its processing results on stdout. Error messages should
+go on stderr.
 
-Invocation of an action will supply parameters as a document on stdin
+If the input object parsed from stdin contains the `output_files` object, the
+action must ALSO write its results, error messages, and exit code in the
+specified files (respectively: `stdout`, `stderr`, and `exitcode`). In this
+case, pxp-agent guarantees that those files will be read only after the
+`exitcode` file is created; the existence of `exitcode` will be used to indicate
```

Comment:
Since the {{exitcode}} file is also used for storing the actual exit code of the module, there is a race condition between creating the file and writing the exit code to it.
To avoid it the each module needs to write its exit status to a file with a different name and then rename it to the name mandated by the ```exitcode``` parameter.
Alternatively we could specify yet another file (other than the ```exitcode``` one) to be used as the indication of the completion of the module execution.